binary_gentoo/internal/cli/tree_diff.py

```diff
@@ -3,12 +3,47 @@
 
 import filecmp
 import os
+import re
 import sys
 from argparse import ArgumentParser
+from typing import Set
 
-from ..reporter import exception_reporting
+from ..reporter import announce_and_check_output, exception_reporting
 from ._parser import add_version_argument_to
 
+_keywords_pattern = re.compile('KEYWORDS="(?P<keywords>[^"]*)"')
+_filename_9999_pattern = re.compile(r'9999(-r[0-9]+)?\.ebuild$')
+
+
+def _replace_special_keywords_for_ebuild(accept_keywords: Set[str],
+                                         ebuild_keywords: Set[str]) -> Set[str]:
+    if '**' in accept_keywords:
+        accept_keywords.remove('**')
+        accept_keywords |= ebuild_keywords
+    elif '*' in accept_keywords:
+        accept_keywords.remove('*')
+        accept_keywords |= {kw for kw in ebuild_keywords if not kw.startswith('~')}
+    elif '~*' in accept_keywords:
+        accept_keywords.remove('~*')
+        accept_keywords |= {kw for kw in ebuild_keywords if kw.startswith('~')}
+    return accept_keywords
```

Comment:
I just noticed a potential bug here. Let me explain what I saw and prove  that (unless mitigated) it might be a real problem:

Superficially, this code reads as if the function takes two sets and produces one new set based on the other two. Sets are passed _by reference_ (rather than _by value_) in Python though, so the `.remove` calls (and apparently also the `|=`) modify the original `accept_keywords` set in the scope outside of this function, they do not work on a copy. Let me prove it to you in IPython:

```python
In [10]: from typing import Set

In [11]: def _replace_special_keywords_for_ebuild(accept_keywords: Set[str],
    ...:                                          ebuild_keywords: Set[str]) -> Set[str]:
    ...:     if '**' in accept_keywords:
    ...:         accept_keywords.remove('**')
    ...:         accept_keywords |= ebuild_keywords
    ...:     elif '*' in accept_keywords:
    ...:         accept_keywords.remove('*')
    ...:         accept_keywords |= {kw for kw in ebuild_keywords if not kw.startswith('~')}
    ...:     elif '~*' in accept_keywords:
    ...:         accept_keywords.remove('~*')
    ...:         accept_keywords |= {kw for kw in ebuild_keywords if kw.startswith('~')}
    ...:     return accept_keywords
    ...: 
    ...: 

In [12]: accept_keywords = {'one', 'two', '**'}

In [13]: ebuild_keywords = {'one', 'three'}

In [14]: replaced = _replace_special_keywords_for_ebuild(accept_keywords, ebuild_keywords)

In [15]: accept_keywords, ebuild_keywords, replaced
Out[15]: ({'one', 'three', 'two'}, {'one', 'three'}, {'one', 'three', 'two'})
```
As you can see, `accept_keywords` no longer contains `'**'` and has `'three'` added now. So when we call `_replace_special_keywords_for_ebuild` next time, results will be unexpected.

For a quick fix we could do e.g. …
- a) `effective_keywords = set(accept_keywords)  # to not affect the outer scope set`
- b) `effective_keywords = copy(accept_keywords)`
  and `from copy import copy`

I'm explicitly ignoring performance here for the moment with the assumption that, the code will run say 10 times rather than 1000 times per runtime and that once (and if) this becomes a bottleneck, we'll need e.g. caching to not compute the same sets over and over, anyway. But let's keep that aside for now and for a later pull request. Okay?

Reply:
Excellent observation. This is not even a potential bug, but a real one. The `_replace_special_keywords_for_ebuild` is called for every ebuild, so basically if the accept_keywords contains "**", it gets set to the keywords of the first ebuild. That is not good.

I have ran some quick performance tests. `set(accept_keywords)` seems to be quicker than `copy(accept_keywords)`, so I'll just include a `set(...)` in that function. 

The copy of a set of 15 short strings takes about 2 microsec on my machine (intel i5), so even if every package in the gentoo repository would have one changed ebuild, the total time spent copying sets would be only 40 ms. That is neglegible. Assuming a system on which you run `binary-gentoo` will be suffuciently performant for the build process itself, I wouldn't invest too much time in performance improvements here.

Reply:
Great, I'm looking forward to the version using `set(..)`